src/commands/node/upgrade.ts

```diff
@@ -0,0 +1,227 @@
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { flags as flagParser } from '@oclif/command';
+import Axios from 'axios';
+import * as fsExtra from 'fs-extra';
+import Listr from 'listr';
+import semver from 'semver';
+import BaseCommand from '../../base';
+import { NETWORK, RELEASE_URL } from '../../utils/constants';
+import { download, extract, validateChecksum } from '../../utils/download';
+import { flags as commonFlags } from '../../utils/flags';
+import { liskTar, liskTarSHA256 } from '../../utils/node/commons';
+import { defaultBackupPath, getConfig } from '../../utils/node/config';
+import {
+	describeApplication,
+	Pm2Env,
+	registerApplication,
+	unRegisterApplication,
+} from '../../utils/node/pm2';
+import { getLatestVersion } from '../../utils/node/release';
+import { exec, ExecResult } from '../../utils/worker-process';
+import StartCommand from './start';
+import StopCommand from './stop';
+
+interface Flags {
+	readonly name: string;
+	readonly version: string;
+}
+
+interface PackageJson {
+	readonly version: string;
+}
+
+const getVersionToUpgrade = async (version: string, network: string) => {
+	if (!version) {
+		const url = `${RELEASE_URL}/${network}/latest.txt`;
+		const latestVersion = await getLatestVersion(url);
+
+		return latestVersion;
+	}
+
+	return version;
+};
+
+const validateVersion = async (
+	network: string,
+	currentVersion: string,
+	toVersion: string,
+): Promise<void> => {
+	if (!semver.valid(toVersion)) {
+		throw new Error(
+			`Upgrade version: ${toVersion} has invalid format, Please refer version from release url: ${RELEASE_URL}/${network}`,
+		);
+	}
+
+	if (semver.lte(toVersion, currentVersion)) {
+		throw new Error(
+			`Upgrade version:${toVersion} should be greater than current version: ${currentVersion}`,
+		);
+	}
+
+	const url = `${RELEASE_URL}/${network}/${toVersion}`;
+	try {
+		await Axios.get(url);
+	} catch (error) {
+		if (error.message === 'Request failed with status code 404') {
+			throw new Error(
+				`Upgrade version: ${toVersion} doesn't exists in ${RELEASE_URL}/${network}`,
+			);
+		}
+		throw new Error(error.message);
+	}
+};
+
+const downloadLisk = async (
+	cacheDir: string,
+	version: string,
+	upgradeVersionURL: string,
+) => {
+	const LISK_RELEASE_PATH = `${cacheDir}/${liskTar(version)}`;
+	const LISK_RELEASE_SHA256_PATH = `${cacheDir}/${liskTarSHA256(version)}`;
+	const liskTarUrl = `${upgradeVersionURL}/${liskTar(version)}`;
+	const liskTarSHA256Url = `${upgradeVersionURL}/${liskTarSHA256(version)}`;
+
+	await download(liskTarUrl, LISK_RELEASE_PATH);
+	await download(liskTarSHA256Url, LISK_RELEASE_SHA256_PATH);
+	await validateChecksum(cacheDir, liskTarSHA256(version));
+};
+
+const backupLisk = async (installDir: string): Promise<void> => {
```

Comment:
maybe move to utils?

Reply:
why? because we will not reuse this function anywhere else, so i dont think moving to utils would be a good idea.

Reply:
I thought this function will be used in migration too?